src/stream_wrap.cc

```diff
@@ -96,20 +96,29 @@ LibuvStreamWrap::LibuvStreamWrap(Environment* env,
 }
 
 
-void LibuvStreamWrap::AddMethods(Environment* env,
-                                 v8::Local<v8::FunctionTemplate> target) {
-  Local<FunctionTemplate> get_write_queue_size =
-      FunctionTemplate::New(env->isolate(),
-                            GetWriteQueueSize,
-                            env->as_external(),
-                            Signature::New(env->isolate(), target));
-  target->PrototypeTemplate()->SetAccessorProperty(
-      env->write_queue_size_string(),
-      get_write_queue_size,
-      Local<FunctionTemplate>(),
-      static_cast<PropertyAttribute>(ReadOnly | DontDelete));
-  env->SetProtoMethod(target, "setBlocking", SetBlocking);
-  StreamBase::AddMethods<LibuvStreamWrap>(env, target);
+Local<FunctionTemplate> LibuvStreamWrap::GetConstructorTemplate(
+    Environment* env) {
+  Local<FunctionTemplate> tmpl = env->libuv_stream_wrap_ctor_template();
+  if (tmpl.IsEmpty()) {
+    tmpl = env->NewFunctionTemplate(nullptr);
+    tmpl->SetClassName(
+        FIXED_ONE_BYTE_STRING(env->isolate(), "LibuvStreamWrap"));
+    tmpl->Inherit(HandleWrap::GetConstructorTemplate(env));
+    Local<FunctionTemplate> get_write_queue_size =
+        FunctionTemplate::New(env->isolate(),
+                              GetWriteQueueSize,
+                              env->as_external(),
+                              Signature::New(env->isolate(), tmpl));
+    tmpl->PrototypeTemplate()->SetAccessorProperty(
+        env->write_queue_size_string(),
+        get_write_queue_size,
+        Local<FunctionTemplate>(),
+        static_cast<PropertyAttribute>(ReadOnly | DontDelete));
+    env->SetProtoMethod(tmpl, "setBlocking", SetBlocking);
+    StreamBase::AddMethods<LibuvStreamWrap>(env, tmpl);
```

Comment:
I’m not sure it’s that simple, because this making this really work would require us to have multiple inheritance in JS – LibuvStreamWrap inherits from both HandleWrap and StreamBase, and… it gets tricky.

The template parameter is currently still necessary, because otherwise the StreamBase methods don’t know at what offset into the C++ object the actual `StreamBase` part begins…

One way that we could get rid of the template parameter and duplicating all the `StreamBase` methods if we were to introduce a second internal field for `StreamBase` JS wrapper objects that points to the `StreamBase*` pointer itself – we already do something pretty similar for the `WriteWrap` and `ShutdownWrap` objects… Any thoughts on that? (tl;dr: Would an extra internal field be worth the simplification of getting rid of the template parameter/casting to `Base` dance?)

Reply:
> Would an extra internal field be worth the simplification of getting rid of the template parameter/casting to Base dance?

Yeah, that sounds better to me (composition over inheritance ;)) AFAICT I don't see it's necessary to add all the methods onto the HandleWrap object itself.

Reply:
Also, I am totally OK with turning that into a composition in another PR

Reply:
I think I prefer to do this in another PR :)